src/Minsk/CodeAnalysis/Emit/Emitter.cs

```diff
@@ -514,6 +521,112 @@ private void EmitBinaryExpression(ILProcessor ilProcessor, BoundBinaryExpression
+
+            switch (nodes.Count)
+            {
+                case 0:
+                    ilProcessor.Emit(OpCodes.Ldstr, string.Empty);
+                    break;
+
+                case 1:
+                    EmitExpression(ilProcessor, nodes[0]);
+                    break;
+
+                case 2:
+                    EmitExpression(ilProcessor, nodes[0]);
+                    EmitExpression(ilProcessor, nodes[1]);
+                    ilProcessor.Emit(OpCodes.Call, _stringConcat2Reference);
+                    break;
+
+                case 3:
+                    EmitExpression(ilProcessor, nodes[0]);
+                    EmitExpression(ilProcessor, nodes[1]);
+                    EmitExpression(ilProcessor, nodes[2]);
+                    ilProcessor.Emit(OpCodes.Call, _stringConcat3Reference);
+                    break;
+
+                case 4:
+                    EmitExpression(ilProcessor, nodes[0]);
+                    EmitExpression(ilProcessor, nodes[1]);
+                    EmitExpression(ilProcessor, nodes[2]);
+                    EmitExpression(ilProcessor, nodes[3]);
+                    ilProcessor.Emit(OpCodes.Call, _stringConcat4Reference);
+                    break;
+
+                default:
+                    ilProcessor.Emit(OpCodes.Ldc_I4, nodes.Count);
+                    ilProcessor.Emit(OpCodes.Newarr, _knownTypes[TypeSymbol.String]);
+
+                    for (var i = 0; i < nodes.Count; i++)
+                    {
+                        ilProcessor.Emit(OpCodes.Dup);
+                        ilProcessor.Emit(OpCodes.Ldc_I4, i);
+                        EmitExpression(ilProcessor, nodes[i]);
+                        ilProcessor.Emit(OpCodes.Stelem_Ref);
+                    }
+
+                    ilProcessor.Emit(OpCodes.Call, _stringConcatArrayReference);
+                    break;
+            }
+
+            static IEnumerable<BoundExpression> Flatten(BoundExpression node)
+            {
+                if (node is BoundBinaryExpression binaryExpression &&
+                    binaryExpression.Op.Kind == BoundBinaryOperatorKind.Addition &&
+                    binaryExpression.Left.Type == TypeSymbol.String &&
+                    binaryExpression.Right.Type == TypeSymbol.String)
+                {
+                    foreach (var result in Flatten(binaryExpression.Left))
+                        yield return result;
+
+                    foreach (var result in Flatten(binaryExpression.Right))
+                        yield return result;
+                }
+                else
+                {
+                    if (node.Type != TypeSymbol.String)
+                        throw new Exception($"Unexpected node type in string concatenation: {node.Type}");
+
+                    yield return node;
+                }
+            }
+
+            static IEnumerable<BoundExpression> FoldConstants(IEnumerable<BoundExpression> nodes)
```

Comment:
Why does this code live here? Shouldn’t this be part of `ConstantFolding`?

*Edit*: I see. You're doing it here because you want to be able collapse trees like this:

```text
     +
    / \
   +   c
  / \
a    b
```

If `b` and `c` are constant, your code will collapse them because they are neighbors in the flattened representation. The constant folder can't do that because this kind of folding would change the tree. This seems like an interesting comment to be added to the code.

Reply:
Yes, that was the intent. I could move it to the `ConstantFolding` class, but it didn't really feel like it belongs there since it rewrites the tree and relies on a list of flattened nodes and can generate its own expression nodes.

If I understand correctly, I should leave it here but add a comment about rewriting `(a + "foo") + "bar"` to `concat(a, "foobar")` ?

Reply:
Yep. I’m not sure generalizing that makes sense yet, so keeping it here is fine. But I’ll ask myself later why it’s here again, so having a comment explaining this would be helpful.

Reply:
Sure, I added a few comments.